khoutsider.py

```diff
@@ -19,97 +53,179 @@ async def download_file(
 
-    url = args.url
+async def download_album(
+    url: str, prefer_flac: bool, output_directory: pathlib.Path
+) -> None:
+    """Top level imperative code for downloading an album."""
 
-    async with aiohttp.ClientSession(raise_for_status=True) as session:
+    retry_options = aiohttp_retry.JitterRetry(attempts=5)
+    async with aiohttp_retry.RetryClient(
+        raise_for_status=True, retry_options=retry_options
+    ) as session:
         try:
             async with session.get(url) as resp:
-                print_if_verbose("Obtained list URL...")
-                album_doc = etree.fromstring(await resp.text(), html_parser)
+                album_doc = etree.fromstring(await resp.text(), HTML_PARSER)
+                LOGGER.info("Obtained list URL for %s", url)
         except aiohttp.ClientError as err:
-            raise
-        print_if_verbose("Obtained URL for ", album_doc.findtext(".//h2"))
-        info_paragraph = etree.tostring(
-            INFO_SELECTOR(album_doc)[0], method="text", encoding="unicode"
-        ).splitlines()
-        for line in info_paragraph:
-            if "Number of Files" in line:
-                track_count = int(line.split(":")[-1])
-                break
-        print_if_verbose(f"{track_count} songs available")
-        async with asyncio.TaskGroup() as tg:
-            for download_page_url in DOWNLOAD_PAGE_SELECTOR(album_doc):
-                tg.create_task(
-                    process_download_page(
-                        urljoin(
-                            url,
-                            min(x.get("href") for x in download_page_url.findall("a")),
-                        ),
-                        session,
-                        html_parser,
+            LOGGER.error("An error occurred in fetching the album at %s: %s", url, err)
+            return
+        album_name = album_doc.findtext(".//h2")
+        if album_name is None:
+            LOGGER.error("Could not find album name for %s", url)
+            return
+        album_directory = output_directory / album_name
+        album_directory.mkdir(exist_ok=True)
+        LOGGER.info("Obtained URL for %s", album_name)
+        try:
+            track_count = get_track_count(album_doc)
+            LOGGER.info("%s songs available for %s", track_count, album_name)
+        except ValueError as err:
+            LOGGER.warning("Could not find album info for %s: %s", album_name, err)
+        try:
+            async with asyncio.TaskGroup() as tg:
+                download_page_urls = DOWNLOAD_PAGE_SELECTOR(album_doc)
+                if len(download_page_urls) == 0:
+                    raise KHOutsiderError(f"No songs found on {url}")
+                for download_page_url in download_page_urls:
+                    tg.create_task(
+                        process_download_page(
+                            urljoin(
+                                url,
+                                # is min necesssary here?
```

Comment:
i had to use a `min` as `findall` returns a set. I don't like it either.

Reply:
Right, but I'm curious if there's a page you ran into that both
1. had multiple `a` tags on that page
2. where min was the proper way to decide between them

In the case where there's only ever one a tag, we can switch from `findall` to `find`, but in the case where there could be multiple, there might need to be some other logic to decide between them.